macro/src/lib.rs

```diff
@@ -32,18 +32,78 @@ fn param_ty(param: &ast::Param) -> syn::Type {
             // not Rust stdlib types (which are not FFI-safe and must be converted)
             prim.get_diplomat_slice_type(ltmt)
         }
-        _ => param.ty.to_syn(),
+        _ => param_ty.to_syn(),
     }
 }
 
-fn param_conversion(param: &ast::Param) -> Option<proc_macro2::TokenStream> {
-    let name = &param.name;
-    match &param.ty {
+fn orig_syn(ty: &ast::TypeName) -> syn::Type {
+    match ty {
+        ast::TypeName::StrReference(lt, encoding, ..) => encoding.get_stdlib_slice_type(lt),
+        ast::TypeName::StrSlice(encoding, ..) => {
+            let inner = encoding.get_stdlib_slice_type(&Some(ast::Lifetime::Anonymous));
+            syn::parse_quote_spanned!(Span::call_site() => &[#inner])
+        }
+        ast::TypeName::PrimitiveSlice(ltmt, primitive, ..) => primitive.get_stdlib_slice_type(ltmt),
+        ast::TypeName::Function(input_types, output_type) => {
+            let input_types: Vec<syn::Type> = input_types
+                .into_iter()
+                .map(|in_ty| orig_syn(in_ty))
+                .collect();
+            let output_type = orig_syn(output_type);
+            syn::parse_quote_spanned!(Span::call_site() => impl Fn(#(#input_types,)*) -> #output_type>)
+        }
+        _ => ty.to_syn(),
+    }
+}
+
+fn param_conversion(
+    name: &ast::Ident,
+    param_type: &ast::TypeName,
+    cast_to: Option<&syn::Type>,
+) -> Option<proc_macro2::TokenStream> {
+    match &param_type {
         // conversion only needed for slices that are specified as Rust types rather than diplomat_runtime types
         ast::TypeName::StrReference(.., StdlibOrDiplomat::Stdlib)
         | ast::TypeName::StrSlice(.., StdlibOrDiplomat::Stdlib)
         | ast::TypeName::PrimitiveSlice(.., StdlibOrDiplomat::Stdlib)
-        | ast::TypeName::Result(..) => Some(quote!(let #name = #name.into();)),
+        | ast::TypeName::Result(..) => Some(if let Some(cast_to) = cast_to {
+            quote!(let #name: #cast_to = #name.into();)
+        } else {
+            quote!(let #name = #name.into();)
+        }),
+        ast::TypeName::Function(in_types, out_type) => {
+            let cb_wrap_ident = &name;
```

Comment:
thought: i don't like how much code this needs but it's understandable.

Reply:
yeah, maybe later we can put it somewhere else to avoid bulking out the macro code

Reply:
yep. i already want to fix param_ty and param_conversion to be `to_ffi_safe_tt()` and `ffi_safe_conversion_tt()` or something

Reply:
ah sweet, that's a good idea